messages/compiler/cpp/cpp_visitor.py

```diff
@@ -0,0 +1,261 @@
+# Concord
+#
+# Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+#
+# This product is licensed to you under the Apache 2.0 license (the "License").
+# You may not use this product except in compliance with the Apache 2.0 License.
+#
+# This product may include a number of subcomponents with separate copyright
+# notices and license terms. Your use of these subcomponents is subject to the
+# terms and conditions of the subcomponent's license, as noted in the LICENSE
+# file.
+
+from visitor import Visitor
+
+
+def struct_start(name, id):
+    return f"""
+struct {name} {{
+  static constexpr uint32_t id = {id};
+
+"""
+
+
+def serialize_start(name):
+    return f"""
+void serialize(std::vector<uint8_t>& output, const {name}& t) {{
+"""
+
+
+def deserialize_start(name):
+    return f"""
+void deserialize(uint8_t*& input, const uint8_t* end, {name}& t) {{
```

Comment:
Is the input mutable? Can it be constant?

Reply:
Input is mutable, otherwise I would have taken it directly by const ref or pointer, or used a const_iterator.  Since deserialization is recursive, we need to move the pointer along the input buffer to know where to continue deserializing from.